# Load & Inspect

## Load & General Inspectation

In [1]:
import pandas as pd
import os
from sklearn.model_selection import train_test_split

In [3]:
data = pd.read_csv('../data/raw/train.csv');

print("data shape:")
print(data.shape)
print("--------------------------------")
print("data types:")
print(data.dtypes)
print("--------------------------------")
print("missing values:")
print(data.isnull().sum())
print("--------------------------------")
print("duplicate rows:")
print(data.duplicated().sum())
print("--------------------------------")
data.head(10)

data shape:
(891, 12)
--------------------------------
data types:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object
--------------------------------
missing values:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64
--------------------------------
duplicate rows:
0
--------------------------------


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


**Observation**

- There are 891 rows and 12 columns including the target.
- PassengerId, Survived, Pclass, SibSp and Parch are **integers**, Name, Sex, Ticket, Cabin and Embarked are **strings**, Age and Fare are **floats**. 
- 19.87% of rows miss the age variable, 77.1% miss the Cabin variable and 0.22% miss the Embarked variable.. 
- There aren't any duplicate row.


## Features Inspectation

### Overview

- As mentioned above **Survived** is the target variable and it is **categorical*.
- **Pclass** represents the passenger's class he payed for in titanic. There are three classes from lowest to highest 3, 2 and 1 (**categorical*).
- **Name** it is a string that represents the name of the passenger, in this study it is not informative to decide wether the passenger will survive or not.
- **Sex** represent the sex of the passenger with two classes female or male (**categorical*).
- **Age** (**numerical*)
- **SibSp** the dataset defines family relations in this way...
Sibling = brother, sister, stepbrother, stepsister
Spouse = husband, wife (mistresses and fiancés were ignored)
(**numerical*)
- **Parch** The dataset defines family relations in this way...
Parent = mother, father
Child = daughter, son, stepdaughter, stepson
Some children travelled only with a nanny, therefore parch=0 for them.
(**numerical*)
- **Ticket** represents the ticket number, it is a string.
- **Fare** represents the passenger's fare, i noticed that when the class is higher this value becomes higher too and when it is lower the value becomes lower (**numerical*).
- **Cabin** the passenger's room inside the titanic during the journey, i noticed that the passengers with class 3 or sometimes 2 have no Cabin value (missing), it is a string (just an hypothesis for now we will confirm that later on).
- **Embarked** Port of Embarkation C = Cherbourg, Q = Queenstown, S = Southampton (**categorical*).
 

#### 1. Categorical Features

In [ ]:
data['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

**Observation**
- With this result we confirmed that the Pclass feature can have three values (1,2 or 3) and there is a significant dominance in class 3 meaning most of passengers payed for class 3.

In [13]:
data['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

**Observation**
- Most of passengers were males.

In [14]:
data['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

**Observation**

- S = Southampton, England — the ship's starting point, where the majority of passengers boarded.
- C = Cherbourg, France — first stop, on the northern French coast.
- Q = Queenstown, Ireland — final stop before heading out into the open Atlantic (modern name: Cobh).

#### 2. Numerical Features

In [15]:
data[['Age', 'Fare', 'SibSp', 'Parch']].describe()

,Age,Fare,SibSp,Parch
count,714.000000,891.000000,891.000000,891.000000
mean,29.699118,32.204208,0.523008,0.381594
std,14.526497,49.693429,1.102743,0.806057
min,0.420000,0.000000,0.000000,0.000000
25%,20.125000,7.910400,0.000000,0.000000
50%,28.000000,14.454200,0.000000,0.000000
75%,38.000000,31.000000,1.000000,0.000000
max,80.000000,512.329200,8.000000,6.000000


**Observation**

- **Age**
    - There are 714 rows out of 891 which confirms what we found earlier there are 177 missing values.
    - mean and median are almost similar (Mean ≈ 29.7, median (50%) = 28) which means that **Age** isn't wildly skewed.
    - min = 0.42 (years ≈ 5 months old) because **Age** is fractional if less than 1.

- **Fare**
    - count = 891 means there are no missing values.
    - mean ≈ 32.2, but median (50%) = 14.45 which means the distribution is right skewed.
    - max = 512.33 and 75th percentile = 31, this returns to the fact that class 3 is dominant and the fare is cheaper therefore 75th percentile had a lower value.

- **SibSp** & **Parch**
    - count = 891 means there are no missing values.
    - SibSp mean ≈ 0.52, median = 0. Parch mean ≈ 0.38, median = 0. The shape for both of them is slightly right skewed.
    - these value indicate that most of passengers travelled alone but there are some outliers max = 8 and max = 6 in row but they are acceptable outliers we may find a familly with many members and all travelled together.

# Split
In this section we will split the train.csv file into X_train, X_val, y_train, y_val x_val and y_val represent the test set, named this way to distinguishe between this test and test.csv file downloaded from kaggle. We will apply the ordinary split 80% of data for training and 20% for testing, and we sill preserve the balance of the two classes (survived and died) along this split.

In [22]:
X = data.drop(columns=['Survived'])
y = data['Survived']

X_train, X_val, y_train, y_val = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

print(y_train.value_counts(normalize=True))
print(y_val.value_counts(normalize=True))

Survived
0    0.616573
1    0.383427
Name: proportion, dtype: float64
Survived
0    0.614525
1    0.385475
Name: proportion, dtype: float64


*Notes*

A few things to notice about this code:

- **stratify= y** forces the split to preserve the same class proportions (~62% died / ~38% survived) in both X_train/y_train and X_val/y_val. Without it, a random split could — just by chance, especially with smaller val sets — land you with a val set that's, say, 70/30 or 50/50, which wouldn't accurately represent the real class balance anymore. That mismatch would make the F1 score on val misleading.
- **random_state= 42** makes the split reproducible; same rows end up in train/val every time you rerun this.
- **test_size= 0.2** 80/20 split, matching what we'd settled on earlier.
- The two print lines let us verify the stratification worked — both should show roughly 62%/38%.

## Save the Files Into interim Folder

In [25]:
output_dir = '../data/interim'  # change this to whatever directory you want
os.makedirs(output_dir, exist_ok=True)  # creates the folder if it doesn't already exist

X_train.to_csv(f'{output_dir}/X_train.csv', index=False)
X_val.to_csv(f'{output_dir}/X_val.csv', index=False)
y_train.to_csv(f'{output_dir}/y_train.csv', index=False)
y_val.to_csv(f'{output_dir}/y_val.csv', index=False)